CoreAudioCaptureSource should have a ref to its CoreAudioSharedUnit

youennf opened the pull request:
#### a051a95261803d47b1f5d47801be6a1b0828400d
<pre>
CoreAudioCaptureSource should have a ref to its CoreAudioSharedUnit
<a href="https://rdar.apple.com/163918367">rdar://163918367</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301847">https://bugs.webkit.org/show_bug.cgi?id=301847</a>

Reviewed by Anne Van Kesteren.

As a preliminary step towards supporting capture of microphone with or without echo cancellation, we make CoreAudioCaptureSource keep a ref to its audio unit.
This reduces the use of CoreAudioSharedUnit::singleton().
There should be no behavioral changes.
Covered by existing tests.

Canonical link: <a href="https://commits.webkit.org/302529@main">https://commits.webkit.org/302529@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a05ec50c726f02960fcb0aeb44f7c6c46a1ec8d0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1637 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136757 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80792 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7e9989d0-6ea5-43fe-a6dc-72ebf2c49255) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131251 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1513 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98535 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66432 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7ac0f4c9-07a8-4969-98ec-bbf4a3ee08be) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132327 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1222 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115877 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79186 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/db633867-1326-4988-99b9-c92458a82ff8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1144 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34009 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80034 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109597 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34508 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139230 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1374 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107060 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1469 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112222 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106904 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27218 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1162 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30742 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54085 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64861 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1315 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1352 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1420 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
```diff
@@ -198,15 +198,26 @@ bool CoreAudioCaptureSourceFactory::shouldAudioCaptureUnitRenderAudio()
-    initializeVolume(CoreAudioSharedUnit::singleton().volume());
+    initializeEchoCancellation(protectedUnit()->enableEchoCancellation());
+    initializeSampleRate(protectedUnit()->sampleRate());
+    initializeVolume(protectedUnit()->volume());
```
If you need a protected method more than once you should use a local variable. (Applies multiple times.)

youennf replied:
Done

annevk commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
```diff
@@ -217,59 +228,59 @@ void CoreAudioCaptureSource::initializeToStartProducingData()
 {
     if (!m_capabilities) {
-        BaseAudioSharedUnit& unit = CoreAudioSharedUnit::singleton();
+        Ref<BaseAudioSharedUnit> unit = m_unit.get();
```
Does this not work?
```suggestion
        Ref unit = m_unit.get();
```

youennf replied:
I migrated the method calls to public to make it work.

annevk commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
```diff
@@ -105,7 +106,11 @@ class CoreAudioCaptureSource : public RealtimeMediaSource, public ThreadSafeRefC
+    Ref<const CoreAudioSharedUnit> protectedUnit() const;
+
     uint32_t m_captureDeviceID { 0 };
+    Ref<CoreAudioSharedUnit> m_unit;
```
Can this be `const` so we can avoid `protectedUnit` completely?

youennf replied:
It can currently but in the future, we will need to support switching from a unit backed by VPIO (only one can run at a time) to a unit that is backed by HAL unit.

annevk commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
```diff
@@ -217,59 +229,62 @@ void CoreAudioCaptureSource::initializeToStartProducingData()
-    CoreAudioSharedUnit::singleton().setEnableEchoCancellation(echoCancellation());
-    CoreAudioSharedUnit::singleton().setSampleRate(sampleRate());
-    CoreAudioSharedUnit::singleton().setVolume(volume());
+    bool shouldReconfigure = echoCancellation() != unit->enableEchoCancellation() || sampleRate() != unit->sampleRate() || volume() != protectedUnit()->volume();
```
```suggestion
    bool shouldReconfigure = echoCancellation() != unit->enableEchoCancellation() || sampleRate() != unit->sampleRate() || volume() != unit->volume();
```

youennf replied:
done

annevk commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
```diff
@@ -217,59 +229,62 @@ void CoreAudioCaptureSource::initializeToStartProducingData()
 
     if (shouldReconfigure)
-        CoreAudioSharedUnit::singleton().reconfigure();
+        protectedUnit()->reconfigure();
```
```suggestion
        unit->reconfigure();
```

youennf replied:
done

annevk commented on Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.cpp:
```diff
@@ -217,59 +229,62 @@ void CoreAudioCaptureSource::initializeToStartProducingData()
         capabilities.setEchoCancellation(m_echoCancellationCapability ? (*m_echoCancellationCapability ? RealtimeMediaSourceCapabilities::EchoCancellation::On : RealtimeMediaSourceCapabilities::EchoCancellation::Off) : RealtimeMediaSourceCapabilities::EchoCancellation::OnOrOff);
         capabilities.setVolume({ 0.0, 1.0 });
-        capabilities.setSampleRate(unit.sampleRateCapacities());
+        capabilities.setSampleRate(unit->sampleRateCapacities());
```
I guess you could use protectedUnit() here as well. Not sure why it was put in a local variable before.

youennf replied:
I directly used m_unit here since we do not need to ref the unit.

annevk reviewed (approved):
Looks okay modulo some nits.

youennf committed:
CoreAudioCaptureSource should have a ref to its CoreAudioSharedUnit
rdar://163918367
https://bugs.webkit.org/show_bug.cgi?id=301847

Reviewed by Anne Van Kesteren.

As a preliminary step towards supporting capture of microphone with or without echo cancellation, we make CoreAudioCaptureSource keep a ref to its audio unit.
This reduces the use of CoreAudioSharedUnit::singleton().
There should be no behavioral changes.
Covered by existing tests.

Canonical link: https://commits.webkit.org/302529@main
